Add ERC: ERC-20/ERC-721 Unified Token Interface (#269)
ethereum/ERCs: ERCS/erc-7629.md

- Defines essential functions and events to support token interactions, conversions, and queries.
- Implements a low gas consumption [ERC-20](./eip-20.md) mode to maintain gas efficiency comparable to typical [ERC-20](./eip-20.md) token transfers.

Compliant contracts MUST implement the following Solidity interface:

```solidity

pragma solidity ^0.8.0;
/**
 * @title ERC-7629 Unify Token Interface
 * @dev This interface defines the ERC-7629 Unify Token, which unifies ERC-721 and ERC-20 assets.
 */
interface IERC7629  is IERC165 {
    // ERC-20 Transfer event
    event ERC20Transfer(
        address indexed from,
        address indexed to,
        uint256 amount
    );

    // ERC-721 Transfer event
    event ERC721Transfer(
        address indexed from,
        address indexed to,
        uint256 indexed tokenId
    );

    // Approval event for ERC-20 and ERC-721
    event Approval(
        address indexed owner,
        address indexed approved,
        uint256 indexed tokenId
    );

    // ApprovalForAll event for ERC-721
    event ApprovalForAll(
        address indexed owner,
        address indexed operator,
        bool approved
    );

>     // ERC-20 to ERC-721 Conversion event
>     event ERC20ToERC721(address indexed to, uint256 amount, uint256 tokenId);
> 
>     // ERC-721 to ERC-20 Conversion event
>     event ERC20ToERC721(address indexed to, uint256 amount, uint256[] tokenIds);

Reviewer: ```suggestion
    // ERC-20 to ERC-721 Conversion event
    event ERC20ToERC721(address indexed to, uint256 amount, uint256 tokenId);

    // ERC-721 to ERC-20 Conversion event
    event ERC721ToERC20(address indexed to, uint256 amount, uint256 tokenId);
```
It seems there are now two `ERC20ToERC721` events defined. It's also preferable to emit a single event per `tokenId` conversion. This approach would enhance the contract's efficiency and make event tracking more straightforward.

Author: In the process of implementation, I've observed that using an array for emitting events during the ERC20 to ERC721 conversion is more suitable

https://github.com/ColorfuLabs/ERC7629

---
eip: 7629
title: Unified Token
description: Unifying ERC-721 and ERC-20 for seamless asset conversion.
author: 0xZeus1111 (@0xZeus1111), Nvuwa (@Nvuwa)
discussions-to: https://ethereum-magicians.org/t/erc-7629-unified-token-protocol/18793
status: Draft
type: Standards Track
category: ERC
created: 2024-02-18

> requires: 20, 721

Reviewer: ```suggestion
requires: 20
```
Requies only the 20 for full compatibility, meaning should inherit IERC20 maybe?

Author: ERC-721 is also required, with ERC-20 being fully compatible and ERC-721 partially compatible.

## Specification

- [ERC-7629](./eip-7629.md) standardizes a token contract that encompasses features from both [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md).
- The token contract supports state transitions between [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md) modes, ensuring smooth conversion and utilization of both liquidity and non-fungibility.
- Defines essential functions and events to support token interactions, conversions, and queries.
- Implements a low gas consumption [ERC-20](./eip-20.md) mode to maintain gas efficiency comparable to typical [ERC-20](./eip-20.md) token transfers.

Compliant contracts MUST implement the following Solidity interface:

```solidity

pragma solidity ^0.8.0;
/**
 * @title ERC-7629 Unify Token Interface
 * @dev This interface defines the ERC-7629 Unify Token, which unifies ERC-721 and ERC-20 assets.
 */

> interface IERC7629 {

Reviewer: ```suggestion
interface IERC7629 is IEC20 {
```
For full ERC20 backward compatibility

    // ERC-20 Transfer event

>     event ERC20Transfer(

Reviewer: ```suggestion
    event Transfer(
```
For full ERC20 backward compatibility

Author: The Transfer event for ERC-20 and ERC-721 is the same. I'd like to retain ERC20Transfer while also adding the Transfer event definition.

Reviewer: I would like to have a clarification on whether this ERC proposal aims to ensure **complete backward compatibility** with ERC20.

    /**
     * @dev Returns the number of decimals used in the token.
     */
    function decimals() external view returns (uint8);

    /**
     * @dev Returns the total supply of the ERC-20 tokens.
     */
    function totalSupply() external view returns (uint256);

    /**
     * @dev Returns the balance of an address for ERC-20 tokens.
     * @param owner The address to query the balance of.
     */
    function balanceOf(address owner) external view returns (uint256);

    /**
     * @dev Returns the total supply of ERC-20 tokens.
     */
    function erc20TotalSupply() external view returns (uint256);

    /**
     * @dev Returns the balance of an address for ERC-20 tokens.
     * @param owner The address to query the balance of.
     */
    function erc20BalanceOf(address owner) external view returns (uint256);

    /**
     * @dev Returns the total supply of ERC-721 tokens.
     */
    function erc721TotalSupply() external view returns (uint256);

    /**
     * @dev Returns the balance of an address for ERC-721 tokens.
     * @param owner The address to query the balance of.
     */
    function erc721BalanceOf(address owner) external view returns (uint256);

> 

Reviewer: ```suggestion
    /**
     * @notice Get the approved address for a single NFT
     * @dev Throws if `tokenId` is not a valid NFT.
     * @param tokenId The NFT to find the approved address for
     * @return The approved address for this NFT, or the zero address if there is none
     */
    function getApproved(uint256 tokenId) external view returns (address);

```
A `getApproved` function is necessary to determine if an NFT has been authorized for a specific address, while this ERC only incorporates the `isApprovedForAll` function.

    /**
     * @dev Checks if an operator is approved for all tokens of a given owner.
     * @param owner The address of the token owner.
     * @param operator The address of the operator to check.
     */
    function isApprovedForAll(
        address owner,
        address operator
    ) external view returns (bool);

    /**
     * @dev Returns the remaining number of tokens that spender will be allowed to spend on behalf of owner.
     * @param owner The address of the token owner.
     * @param spender The address of the spender.
     */
    function allowance(
        address owner,
        address spender
    ) external view returns (uint256);

    /**
     * @dev Returns the array of ERC-721 token IDs owned by a specific address.
     * @param owner The address to query the tokens of.
     */
    function owned(address owner) external view returns (uint256[] memory);

    /**
     * @dev Returns the address that owns a specific ERC-721 token.
     * @param id The token ID.
     */

>     function ownerOf(uint256 id) external view returns (address erc721Owner);

Reviewer: ```suggestion
    function ownerOf(uint256 tokenId) external view returns (address erc721Owner);
```
To avoid ambitious, all ERC721 related `id` should be consistently labelled as `tokenId`

    /**
     * @dev Returns the URI for a specific ERC-721 token.
     * @param id The token ID.
     */

>     function tokenURI(uint256 id) external view returns (string memory);

Reviewer: ```suggestion
    function tokenURI(uint256 tokenId) external view returns (string memory);
```
To avoid ambitious, all ERC721 related `id` should be consistently labelled as `tokenId`

    /**
     * @dev Approve or disapprove the operator to spend or transfer all of the sender's tokens.
     * @param spender The address of the spender.
     * @param amountOrId The amount of ERC-20 tokens or ID of ERC-721 tokens.
     */
    function approve(
        address spender,
        uint256 amountOrId

>     ) external;

Reviewer: ```suggestion
    ) external returns (bool);
```
For full ERC20 backward compatibility

    /**
     * @dev Approve the spender to spend the specified amount of ERC-20 tokens.
     * @param spender The address of the spender.
     * @param value The amount of ERC-20 tokens to approve.
     */
    function erc20Approve(
        address spender,
        uint256 value
    ) external returns (bool);

    /**
     * @dev Approve the operator to spend the specified ERC-721 token.
     * @param spender The address of the spender.
     * @param id The ID of the ERC-721 token to approve.
     */

>     function erc721Approve(address spender, uint256 id) external;

Reviewer: ```suggestion
    function erc721Approve(address spender, uint256 tokenId) external;
```
To avoid ambitious, all ERC721 related `id` should be consistently labelled as `tokenId`

>     function erc721Approve(address spender, uint256 tokenId) external;

Reviewer: ```suggestion
    function erc721Approve(address spender, uint256 tokenId) external payable;
```
From the original `approve` function of ERC721:
`function approve(address _approved, uint256 _tokenId) external payable;`

    /**
     * @dev Transfer ERC-20 tokens or ERC-721 token from one address to another.
     * @param from The address to transfer ERC-20 tokens or ERC-721 token from.
     * @param to The address to transfer ERC-20 tokens or ERC-721 token to.
     * @param amountOrId The amount of ERC-20 tokens or ID of ERC-721 tokens to transfer.
     */
    function transferFrom(
        address from,
        address to,
        uint256 amountOrId
    ) external returns (bool);

    /**
     * @dev Transfer ERC-20 tokens from one address to another.
     * @param from The address to transfer ERC-20 tokens from.
     * @param to The address to transfer ERC-20 tokens to.
     * @param value The amount of ERC-20 tokens to transfer.
     */
    function erc20TransferFrom(
        address from,
        address to,
        uint256 value
    ) external returns (bool);

    /**
     * @dev Transfer ERC-721 token from one address to another.
     * @param from The address to transfer ERC-721 token from.
     * @param to The address to transfer ERC-721 token to.
     * @param id The ID of the ERC-721 token to transfer.
     */
    function erc721TransferFrom(
        address from,
        address to,

>         uint256 id

Reviewer: ```suggestion
        uint256 tokenId
```
To avoid ambitious, all ERC721 related `id` should be consistently labelled as `tokenId`

>     ) external;

Reviewer: ```suggestion
    ) external payable;
```
From the original `transferForm` function of ERC721:
`function transferFrom(address _from, address _to, uint256 _tokenId) external payable;`

> 

Reviewer: ```suggestion
    /**
     * @notice Transfers the ownership of an NFT from one address to another address
     * @dev Throws unless `msg.sender` is the current owner, an authorized
     *  operator, or the approved address for this NFT. Throws if `_rom` is
     *  not the current owner. Throws if `_to` is the zero address. Throws if
     *  `tokenId` is not a valid NFT. When transfer is complete, this function
     *  checks if `to` is a smart contract (code size > 0). If so, it calls
     *  `onERC721Received` on `to` and throws if the return value is not
     *  `bytes4(keccak256("onERC721Received(address,address,uint256,bytes)"))`.
     * @param from The current owner of the NFT
     * @param to The new owner
     * @param tokenId The NFT to transfer
     * @param data Additional data with no specified format, sent in call to `to`
     */
    function safeTransferFrom(address from, address to, uint256 tokenId, bytes data) external payable;

    /**
     * @notice Transfers the ownership of an NFT from one address to another address
     * @dev This works identically to the other function with an extra data parameter,
     *  except this function just sets data to "".
     * @param from The current owner of the NFT
     * @param to The new owner
     * @param tokenId The NFT to transfer
     */
    function safeTransferFrom(address from, address to, uint256 tokenId) external payable;

```
The `safeTransferFrom` functions of ERC721 might be subject to review.

---
eip: 7629
title: Unified Token
description: Unifying ERC-721 and ERC-20 for seamless asset conversion.
author: 0xZeus1111 (@0xZeus1111), Nvuwa (@Nvuwa)
discussions-to: https://ethereum-magicians.org/t/erc-7629-unified-token/18793
status: Draft
type: Standards Track
category: ERC
created: 2024-02-18
requires: 20, 165, 721
---

## Abstract

> [ERC-7629](./eip-7629.md), known as the Unified Token, introduces a comprehensive protocol unifying the characteristics of [ERC-721](./eip-721.md) and [ERC-20](./eip-20.md) tokens within the Ethereum ecosystem. This standard seamlessly integrates the liquidity features of [ERC-20](./eip-20.md) with the non-fungible nature of [ERC-721](./eip-721.md), enabling frictionless conversion between these asset types. [ERC-7629](./eip-7629.md) offers a multifunctional solution, providing developers and users with the flexibility to leverage both liquidity and non-fungibility in a unified token framework.

Reviewer 2: ```suggestion
This proposal introduces the Unified Token, a comprehensive protocol unifying the characteristics of [ERC-721](./eip-721.md) and [ERC-20](./eip-20.md) tokens within the Ethereum ecosystem. This standard seamlessly integrates the liquidity features of [ERC-20](./eip-20.md) with the non-fungible nature of [ERC-721](./eip-721.md), enabling frictionless conversion between these asset types. [ERC-7629](./eip-7629.md) offers a multifunctional solution, providing developers and users with the flexibility to leverage both liquidity and non-fungibility in a unified token framework.
```

     * @dev This works identically to the other function with an extra data parameter,
     *  except this function just sets data to "".
     * @param from The current owner of the NFT
     * @param to The new owner
     * @param tokenId The NFT to transfer
     */
    function safeTransferFrom(address from, address to, uint256 tokenId) external payable;

    /**
     * @dev Transfer ERC-20 tokens to an address.
     * @param to The address to transfer ERC-20 tokens to.
     * @param amount The amount of ERC-20 tokens to transfer.
     */
    function transfer(address to, uint256 amount) external returns (bool);

    /**
     * @dev Retrieves the unit value associated with the token.
     * @return The unit value.
     */
    function getUnit() external view returns (uint256);

    /**
     * @dev Converts ERC-721 token to ERC-20 tokens.
     * @param tokenId The unique identifier of the ERC-721 token.
     */
    function erc721ToERC20(uint256 tokenId) external;

    /**
     * @dev Converts ERC-20 tokens to an ERC-721 token.
     * @param amount The amount of ERC-20 tokens to convert.
     */
    function erc20ToERC721(uint256 amount) external;
}

```
## Rationale

> The rationale for [ERC-7629](./eip-7629.md) lies in its ability to harmonize liquidity and non-fungibility, filling a crucial gap in the existing token standards. By unifying [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md) features, [ERC-7629](./eip-7629.md) enhances user flexibility, allowing for efficient conversion between these asset types. The protocol is designed to provide a standardized and versatile solution for developers, promoting a more streamlined development process. With smooth transitions and low gas consumption, [ERC-7629](./eip-7629.md) aims to be a robust standard that not only simplifies token interactions but also fosters broader adoption within the Ethereum ecosystem.

Reviewer 2: Your Rationale section is not explaining any decisions made within the document. Instead, you seem to have written more motivation-style content here. This section is meant to explain decisions you've made in your design, not justify the document as a whole. My favourite analogy is:

> Motivation: We need to build a shed because...
> Rationale: We painted the shed red because...

### Compatibility Points

The primary compatibility point lies in the discrepancy between [ERC-20](./eip-20.md)'s balanceOf and [ERC-721](./eip-721.md)'s balanceOf functionalities. Developers accustomed to the specific balance query methods in each standard may face challenges when transitioning to [ERC-7629](./eip-7629.md).

### Proposed Solutions

Dual Balance Query Functions:

Introduce two distinct functions, `erc20BalanceOf` and `erc721TotalSupply`, to align with the conventions of [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md), respectively. Developers can choose the function based on the token type they are working with.

## Test Cases

## Reference Implementation

- [ERC-7629](./eip-7629.md) standard undergoes thorough testing to ensure consistent and accurate behavior in both [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md) modes.
- Detailed documentation and sample code are provided to assist developers in understanding and implementing the [ERC-7629](./eip-7629.md) standard.

> `https://github.com/ColorfuLabs/ERC7629`

Reviewer 2: ```suggestion
```

Please remove this external link. If you'd like to include a reference implementation, you may do so in the `assets` directory.

---
eip: 7629
title: Unified Token
description: Unifying ERC-721 and ERC-20 for seamless asset conversion.
author: 0xZeus1111 (@0xZeus1111), Nvuwa (@Nvuwa)
discussions-to: https://ethereum-magicians.org/t/erc-7629-unified-token/18793
status: Draft
type: Standards Track
category: ERC
created: 2024-02-18

> requires: 20, 721

Reviewer 3: As IERC7629 is going to create a new interface, better based on IERC165.

Author: Thank you for your advice. I will add the IERC165 interface to IERC7629.

## Specification

- [ERC-7629](./eip-7629.md) standardizes a token contract that encompasses features from both [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md).
- The token contract supports state transitions between [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md) modes, ensuring smooth conversion and utilization of both liquidity and non-fungibility.
- Defines essential functions and events to support token interactions, conversions, and queries.
- Implements a low gas consumption [ERC-20](./eip-20.md) mode to maintain gas efficiency comparable to typical [ERC-20](./eip-20.md) token transfers.

Compliant contracts MUST implement the following Solidity interface:

```solidity

pragma solidity ^0.8.0;
/**
 * @title ERC-7629 Unify Token Interface
 * @dev This interface defines the ERC-7629 Unify Token, which unifies ERC-721 and ERC-20 assets.
 */

> interface IERC7629 {

Reviewer 3: ```
interface IERC7629 /* is ERC165 */ {
```

- The token contract supports state transitions between [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md) modes, ensuring smooth conversion and utilization of both liquidity and non-fungibility.
- Defines essential functions and events to support token interactions, conversions, and queries.
- Implements a low gas consumption [ERC-20](./eip-20.md) mode to maintain gas efficiency comparable to typical [ERC-20](./eip-20.md) token transfers.

Compliant contracts MUST implement the following Solidity interface:

```solidity

pragma solidity ^0.8.0;
/**
 * @title ERC-7629 Unify Token Interface
 * @dev This interface defines the ERC-7629 Unify Token, which unifies ERC-721 and ERC-20 assets.
 */
interface IERC7629 is IERC165{
    // ERC-20 Transfer event
    event ERC20Transfer(
        address indexed from,
        address indexed to,
        uint256 amount
    );

    // ERC-721 Transfer event
    event ERC721Transfer(
        address indexed from,
        address indexed to,
        uint256 indexed tokenId
    );

    // Approval event for ERC-20 and ERC-721
    event Approval(
        address indexed owner,
        address indexed approved,
        uint256 indexed tokenId

>     );

Reviewer: I've reviewed the proposed event signature for the unified Approval event intended to serve both ERC-20 and ERC-721 tokens, while feeling concerned about the potential ambiguity the current design might introduce. Specifically, it's unclear how to differentiate whether the third parameter, tokenId, represents an ERC-721 token's unique identifier or an ERC-20 token's value. In ERC-20 contexts, this parameter would typically denote a quantity, whereas in ERC-721 contexts, it uniquely identifies a token.

The proposed this proposal introduces a challenge in terms of backward compatibility due to the distinct balance query mechanisms utilized by [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md) standards. [ERC-20](./eip-20.md) employs `balanceOf` to check an account's token balance, while [ERC-721](./eip-721.md) uses `balanceOf` to inquire about the quantity of tokens owned by an account. To reconcile these differences, the ERC must consider providing either two separate functions catering to each standard or adopting a more generalized approach.

### Compatibility Points

The primary compatibility point lies in the discrepancy between [ERC-20](./eip-20.md)'s balanceOf and [ERC-721](./eip-721.md)'s balanceOf functionalities. Developers accustomed to the specific balance query methods in each standard may face challenges when transitioning to this proposal.

### Proposed Solutions

Dual Balance Query Functions:

Introduce two distinct functions, `erc20BalanceOf` and `erc721TotalSupply`, to align with the conventions of [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md), respectively. Developers can choose the function based on the token type they are working with.

> ## Test Cases
> 
> 
> ## Reference Implementation
> 
> - this proposal standard undergoes thorough testing to ensure consistent and accurate behavior in both [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md) modes.
> - Detailed documentation and sample code are provided to assist developers in understanding and implementing the this proposal standard.

Reviewer 2: ```suggestion
```

These sections are optional; you can remove them until you have content.

---
eip: 7629
title: ERC-20/721 Unified Token Interface

> description: Unified ERC-20/ERC-721 interface simplifies token management.

Reviewer 2: You still don't really say what "unified" means. Does it mean you can convert your ERC-721 into ERC-20 tokens on the fly? Is this defining a subset of functions common to both ERC-721 and ERC-20? You _need_ to give the reader more information here.

## Abstract

> This proposal introduces a protocol that provides a unified interfacefor managing both [ERC-20](./eip-20.md) fungible tokens and [ERC-721](./eip-721.md) non-fungible tokens (NFTs) on the Ethereum blockchain. The protocol defines a set of common functions, including transferFrom, mint, burn, balanceOf, and ownerOf, which are applicable to both token types. By standardizing these functions, developers can interact with both [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md) tokens using a single, cohesive interface, simplifying integration and enhancing interoperability within decentralized applications (DApps).

Reviewer 2: ```suggestion
This proposal introduces a protocol that provides a unified interface for managing both [ERC-20](./eip-20.md) fungible tokens and [ERC-721](./eip-721.md) non-fungible tokens (NFTs) on the Ethereum blockchain. The protocol defines a set of common functions, including `transferFrom`, `mint`, `burn`, `balanceOf`, and `ownerOf`, which are applicable to both token types. By standardizing these functions, developers can interact with both [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md) tokens using a single, cohesive interface, simplifying integration and enhancing interoperability within decentralized applications (DApps).
```
